packages/edit-navigation/src/components/menu-locations-editor/index.js

```diff
@@ -1,3 +1,104 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import {
+	SelectControl,
+	Button,
+	Panel,
+	PanelBody,
+	Spinner,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import useMenuLocations from './use-menu-locations';
+
 export default function MenuLocationsEditor() {
-	return <>Menu locations editor</>;
+	const menus = useSelect( ( select ) => select( 'core' ).getMenus() );
+
+	const [
+		menuLocations,
+		saveMenuLocations,
+		assignMenuToLocation,
+	] = useMenuLocations();
+
+	if ( ! menus || ! menuLocations ) {
+		return <Spinner />;
+	}
+
+	const menuSelectControlOptions = [
+		{ value: 0, label: __( '— Select a Menu —' ) },
+		...menus.map( ( { id, name } ) => ( {
+			value: id,
+			label: name,
+		} ) ),
+	];
+
+	if ( menuLocations.length === 0 ) {
+		return (
+			<Panel className="edit-navigation-menu-editor__panel">
+				<PanelBody title={ __( 'Menu locations' ) }>
+					<p>{ __( 'There are no available menu locations' ) }</p>
+				</PanelBody>
+			</Panel>
+		);
+	}
+
+	if ( menus.length === 0 ) {
+		return (
+			<Panel className="edit-navigation-menu-editor__panel">
+				<PanelBody title={ __( 'Menu locations' ) }>
+					<p>{ __( 'There are no available menus' ) }</p>
+				</PanelBody>
+			</Panel>
+		);
+	}
+
+	return (
+		<Panel className="edit-navigation-menu-editor__panel">
+			<PanelBody title={ __( 'Menu locations' ) }>
+				<form
+					onSubmit={ ( event ) => {
+						event.preventDefault();
+						saveMenuLocations();
+					} }
+				>
+					<table>
+						<thead>
+							<tr>
+								<th scope="col">{ __( 'Theme Location' ) }</th>
+								<th scope="col">{ __( 'Assigned Menu' ) }</th>
+							</tr>
+						</thead>
+						<tbody>
+							{ menuLocations.map( ( location ) => (
+								<tr key={ location.name }>
+									<td>{ location.description }</td>
+									<td>
+										<SelectControl
+											options={ menuSelectControlOptions }
+											value={ location.menu }
+											onChange={ ( newMenuId ) => {
+												location.menu = newMenuId;
```

Comment:
I wouldn't expect that this line has any effect as it's updating an object that exists in React component state using direct mutation instead of `setState`.

Probably this needs to be a call to `setMenuLocations()` in `assignMenuToLocation` instead.

Reply:
It has the effect of setting the new selected value on the select component. Doesn't `location` in this context exist in the scope of the `map` callback only? I think it does not affect any state. If I update the `menuLocations` state, we'll see a re-render and a spinner.

Reply:
Technically this works since `menuLocations` is only ever set once, but in general mutating the state is an anti-pattern. React state should be immutable. If someone else picks up this code a few months from now and call `setMenuLocations` again for some reason, they would override the information you're storing here. It would be pretty hard to debug what went wrong as direct state mutations are not something you expect. Also, from React docs:

> Never mutate this.state directly, as calling setState() afterwards may replace the mutation you made. Treat this.state as if it were immutable.

I think the solution involves updating the state using `setMenuLocations` as @noisysocks recommended.

Reply:
> If I update the menuLocations state, we'll see a re-render and a spinner.

If I'm reading it right, the spinner will only be there if menuLocations are empty. So if you update it to a non-null value, I think it should be okay.

Reply:
I've updated the PR so that we're setting and updating menuLocations when assigning menus to locations. This reduced the code a bit as well.